### ISSUE TITLE
Allow overwriting of slug name chosen for resources

### ISSUE THREAD

**Issue by kdawgwilk** [vapor/vapor#691]
# Feature, Enhancement, or Optimization

# Overwrite slug name for resources

* Author(s): [Kaden Wilkinson](https://github.com/kdawgwilk)

## Introduction

I would like to choose what the slug name is used when creating resource routes so that I can extend those routes easier.

## Motivation

Here is a sample of what I am trying to accomplish:

```swift
let drop = Droplet()

let api: RouteGroup  = drop.grouped("api")
let v1: RouteGroup = api.grouped("v1")

// /users
v1.resource("users", UserController())

let user: RouteGroup = v1.grouped("users/:user_id")

// /users/:user_id/groups
user.resource("groups", GroupController())

let group: RouteGroup = v1.grouped("users/:user_id/groups/:group_id")

// /users/:user_id/groups/:group_id/tasks
group.resource("tasks", TaskController())

// /users/:user_id/groups/:group_id/rewards
group.resource("rewards", RewardController())
```

but I keep getting these warnings in the console:

```
[WARNING] Mismatched Slugs:
Attempted to overwrite :id with :user_id
Please use the same slug name for all routes on shared branch
```

## Proposed solution

Currently the slug name is hard coded in the function for creating a new resource.  I propose to move this to a default parameter so that it can be overridden if needed.

## Code snippets

This would allow you to then say something similar to to this:

```swift
let drop = Droplet()

let api: RouteGroup  = drop.grouped("api")
let v1: RouteGroup = api.grouped("v1")

// /users
v1.resource("users", ":user_id", UserController())

let user: RouteGroup = v1.grouped("users/:user_id")

// /users/:user_id/groups
user.resource("groups", ":group_id", GroupController())

let group: RouteGroup = v1.grouped("users/:user_id/groups/:group_id")

// /users/:user_id/groups/:group_id/tasks
group.resource("tasks", ":task_id", TaskController())

// /users/:user_id/groups/:group_id/rewards
group.resource("rewards", ":reward_id", RewardController())
```

## Impact

By using a default parameter it should not effect existing code but support greater flexibility.

## Alternatives considered

This was inspired as a workaround by my struggling with the previous issue #676

**Comment by LoganWright** [comment]
@kdawgwilk this looks good, I think the trick here is that the controller would need to know about `task_id` somehow. Don't have a solution yet, we could consider passing it in or something perhaps, just laying it here so we know the issue.

**Comment by kdawgwilk** [comment]
@LoganWright what I did to fix the controller problem was made a TaskResource protocol that had a default implementation that extracted the task_id and found the task so I could access it in the controller.  The problem still is that I had to hard code the "task_id" string in the protocol

**Comment by kdawgwilk** [comment]
This issue has been mostly addressed in #693 which has been merged! 👏
